### PR TITLE
Disable agent on early Java 8 versions

### PR DESCRIPTION
**Pull request by SylvainJuge**
# What does this PR do?

Early Java 8 versions are known to have bugs that are triggered when used with an agent doing instrumentation. Of course, those flaws might not be visible without the agent, thus agent becomes the obvious culprit whereas in practice we can't really work around those.

As a result, using our agent with one of those JVMs will likely result in:
- a crashed JVM
- likely some angry ops, because killing, even if it's a computer process is a bad thing.
- a lot of back & forth with user/customer and support trying to debug a core dump
- unmet expectations: Java 8 is officially supported.

## Checklist
- [x] My code follows the [style guidelines of this project](CONTRIBUTING.md#java-language-formatting-guidelines)
- [x] I have rebased my changes on top of the latest master branch
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing [**unit** tests](https://github.com/elastic/apm-agent-java/blob/master/CONTRIBUTING.md#testing) pass locally with my changes
- [x] I have updated [CHANGELOG.asciidoc](CHANGELOG.asciidoc)

## Author's Checklist
- [x]  run integration tests to make sure none of our docker images is using one of those JVMs
- [x]  test manually on working / not buggy versions

### CODE REVIEW

File: apm-agent-core/src/main/java/co/elastic/apm/agent/bci/AgentMain.java [hunk 1]
```diff
@@ -90,6 +98,43 @@ public synchronized static void init(String agentArguments, Instrumentation inst
...
+            return false;
+        } else if (major == '7' || major > '8') {
+            return true;
+        } else {
+            // java 8
+            int updateIndex = javaVersion.lastIndexOf("_");
+            if (updateIndex <= 0) {
+                return false;
+            } else {
+                String updateVersion = javaVersion.substring(updateIndex + 1);
```
Comment (eyalkoren) [inline apm-agent-core/src/main/java/co/elastic/apm/agent/bci/AgentMain.java]: One of our Tomcat tests fails due to the version called: `1.8.0_72-internal` 🙁

**Comment by eyalkoren** [comment]
Indeed, one of the WebSphere tests fails with `[2020-03-05T07:06:27.540Z] [websphere] Failed ot start agent - JVM version not supported: 1.8.0`

File: apm-agent-core/src/main/java/co/elastic/apm/agent/bci/AgentMain.java [hunk 2]
```diff
@@ -72,8 +72,16 @@ public synchronized static void init(String agentArguments, Instrumentation inst
             // for example, Spring Boot restarts the application in dev mode
             return;
         }
+
+        String javaVersion = System.getProperty("java.version");
+        if (!isJavaVersionSupported(javaVersion)) {
+            // gracefully abort agent startup is better than unexpected failure down the road
+            System.err.println(String.format("Failed ot start agent - JVM version not supported: %s", javaVersion));
```
Comment (eyalkoren) [inline apm-agent-core/src/main/java/co/elastic/apm/agent/bci/AgentMain.java]:
```suggestion
            System.err.println(String.format("Failed to start agent - JVM version not supported: %s", javaVersion));
```

File: apm-agent-core/src/main/java/co/elastic/apm/agent/bci/AgentMain.java (hunk 1 above)
Comment (SylvainJuge) [reply apm-agent-core/src/main/java/co/elastic/apm/agent/bci/AgentMain.java]: fixed in the last commit.

**Comment by SylvainJuge** [comment]
> We probably want to do that vendor-specific. Not sure about compatibility between Oracle and OpenJDK, but I assume IBM's J9 doesn't have to be under same restriction.

That's a very good question, I'm fine to disable this check for IBM JVMs, but only apply it to openjdk/oracle releases that are known to be problematic.

File: apm-agent-core/src/main/java/co/elastic/apm/agent/bci/AgentMain.java [hunk 3]
```diff
@@ -99,43 +105,52 @@ public synchronized static void init(String agentArguments, Instrumentation inst
...
         if (major < '7') {
             // given code is compiled with java 7, this one is unlikely in practice
             return false;
         } else if (major == '7' || major > '8') {
             return true;
+        } else if (!vmName.contains("HotSpot(TM)")) {
+            // non-hotspot JVMs are not concerned (yet)
+            return true;
         } else {
             // java 8
```
Comment (eyalkoren) [inline apm-agent-core/src/main/java/co/elastic/apm/agent/bci/AgentMain.java]:
```suggestion
            // HotSpot 8
```

**Pull request merged**